src/track/track.cpp

```diff
@@ -431,9 +438,52 @@ bool Track::trySetAndLockBeats(mixxx::BeatsPointer pBeats) {
 }
 
 bool Track::setBeatsWhileLocked(mixxx::BeatsPointer pBeats) {
+    qWarning() << "     .";
+    qWarning().noquote() << "     setBeatsWhileLocked "
+                         << "old beats:" << QString(m_pBeats != nullptr ? "X" : "-")
+                         << "| new beats:" << QString(pBeats != nullptr ? "X" : "-");
     if (m_pBeats == pBeats) {
+        qWarning() << "       old == new, return";
+        qWarning() << "     .";
         return false;
     }
+    if (m_undoingBeatsChange) {
+        qWarning() << "       (currently undoing, don't rec)";
+    }
+    if (m_pBeats == nullptr) {
+        qWarning() << "       (currently null beats, don't rec)";
+    }
+
+    // Don't add null beats to the undo stack. Happens when beats are deserialized,
+    // e.g. when opening the track menu.
+    // Don't add beats to stack which we're about to undo.
+    mixxx::Duration time = mixxx::Time::elapsed();
```

Comment:
I stumbled here for a short moment about "elapsed() since what?".
Maybe we should rename it to `mixxx::Time::sinceStart()` or such. 

In addition we may want to use a "PerformanceTimer" object instead of implementing it inline. 
Using `start()` `restart()` and `elapsed()`

Reply:
Did you consider this?

Reply:
I'm considering this, yes.
Pro argument for the current solution is that we don't create an extra timer object, we simply use the Mixxx time.
A performance timer would be running all the time **for each track we touch**, even if the user doesn't change the beats (apart from initial load/analyse). A regular one-shot timer (auto-stop) might also work (delta is 800 ms).
Hence I don't really see the necessity to change this.

Reply:
First of all, I have to a strong demand here. We can keep it as it is. 

But let's give some insights: 
The PerformanceTimer object not "running" in terms of CPU. It is just a thin wrapper around a single member variable that stores the start time. See: https://github.com/mixxxdj/mixxx/blob/7111b70e3fa346b67bab3f275b8cce037ba6274b/src/util/performancetimer.h#L37
So basically your current solution wrapped by an API.

Reply:
Oh, sorry, should have looked closer.
I'll try that!